### ISSUE TITLE
CI for commit title guidelines adherence

### ISSUE THREAD

**Issue by Atemu** [NixOS/nixpkgs#281573]
## Issue description

Lots of new contributors (and even some old hands) create commit titles that do not conform with [our standards](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions).

We should have a CI check for that which explains how to do it correctly on failure such that reviewers don't have to.

Some loose check requirements off the top of my head:

- Starts with lower case
- Starts with `<component>: `
  - Commits touching `nixos` should be `nixos/<module>: `
  - Commits touching `lib` should be `lib.<something>: `
  - Commits touching a package in `pkgs/by-name/pk/pkg/` should be `<pkg>: `
  - Commits touching a package not in `pkgs/by-name/` should be `<string>: `
  - Commits touching `doc` should be `doc: `
- Commits touching multiple areas or sub-areas may conform with any one of the areas they're touching
- Lower case after `:`
- No notifications in draft PRs (to keep the ability for quicker prototyping)
- No notifications for `fixup!` `squash!` etc. commit titles to keep the ability to use the GH "commit suggestion" feature (though merge should ideally be blocked if any of these are still present)

(Please comment with more if there's something I hadn't considered.)

There are valid commit titles which do not start with `<component>: ` but they're very rare IME. I can only think of release related commits off the top of my head.  
In those cases, I think it'd be fine to simply merge despite CI. The worst thing that happens is that the person gets a github notification about a failed CI check which I think is acceptable given how rare it should be.  
Please comment if there are cases more common than that which require non-conformant commit titles.

Another thing we could automate here is checks against merging master into the PR branch rather than rebasing (disabled if source is staging-next).

cc people who might like this idea: @piegamesde @infinisil @zimbatm

**Comment by infinisil** [comment]
I'm in favor of more automation, but only for things that actually make sense and are beneficial. Let's be very specific about what we want to enforce and why.

> Starts with <component>:

What's the reason behind wanting to do this?
- The fact that ofborg needs to [use the commit messages](https://github.com/NixOS/ofborg#automatic-building) to infer what to build is just a hack imo, we could implement something better.
- Related is [conventional commits](https://www.conventionalcommits.org/), but the main purpose of that is automated SemVer tooling, which is not relevant for us.
- I guess this might be mainly for a clean and searchable history, since you can just `grep` for titles to find stuff. That seems like a decent reason to do this. However I also think this is a symptom of poor tooling, e.g. if we had an easy way of looking up which commits changed packages (which btw could be much more informative than just version information), we wouldn't really need this.

> Lower case after :

This seems entirely pointless to me, I can't see a reason to do that.

**Comment by zimbatm** [comment]
Before dismissing an idea, we can take a minute. I don't think that Ofborg will be rewritten tomorrow. As long as we have those conventions in place, and if Atemu is motivated to implement those checks, then why not.

My only concern is if it adds too much friction to the contribution process. I would probably start with a subset of those checks and see if it's viable. Maybe add the other ones as warnings.

**Comment by Atemu** [comment]
> What's the reason behind wanting to do this?

To only have one possible way to format commit messages.

The primary thing this avoids is bikeshedding/nitpicking and other noise not actually related to the primary purpose of a given PR.

What prompted me to open this is https://github.com/NixOS/nixpkgs/pull/274248. Fully understandable patterns from both "sides" but it just took way too much back and forth to get the PR into a mergeable state because of this.  
If we had a CI check, the author would have gotten instant feedback and with no confusion whether their commit now has the correct name either as there'd be a clear marker.

Being able to do some automatic analysis is also neat but more of a secondary goal.

Also note that, while such a common format can aid automation, it can also aid humans. E.g. you know there was a commit that changed something about a package for its module a year ago and you need to find that commit again.  
If I saw a commit titled `nixos/pkg: ...` in the package's file's history, that'd be the first I'd look at.  
Yes, tooling could help in some of these cases but having human-interpretable commit messages as a simple fall-back would still be handy even if some of that tooling existed. (I also don't see how there could ever be performant Nix tooling that could help navigate the git history.)

> This seems entirely pointless to me, I can't see a reason to do that.

Also stylistic unification.

I don't actually care which way we're doing it but I only want one of them as it makes things easier to parse at a glance. As one of the authors of RFC 166, I'm sure you understand ;)

I chose lower case simply because it's the predominant style here and also used in other projects such as the Linux kernel.

**Comment by piegamesde** [comment]
While I appreciate the effort and would like to see this, I think it should not be too strict or pedantic. There is a balance to hit there. For example, I personally wouldn't care that much about casing. Also, there still are many cases where free-form commit titles are a thing, and people shouldn't have to contort themselves around the CI for these.

I think the best way would be to start with a small set of rules but which don't make CI red, and then expand as new edge cases are found until it handles most PRs sufficiently well.

**Comment by Atemu** [comment]
> Also, there still are many cases where free-form commit titles are a thing, and people shouldn't have to contort themselves around the CI for these.

Absolutely. I've already identified the release process as one of these. Part of the reason for opening this issue is finding out whether others know of more such cases.

> I think the best way would be to start with a small set of rules but which don't make CI red, and then expand as new edge cases are found until it handles most PRs sufficiently well.

I was thinking having a check for `<anything>: <anything>` would be a good first start. This should cover likely >90% of wrong commit titles.

This could then be expanded by the more precise checks (including case checks) as warnings at first until we get a good feeling about them and then have those turn into annoying red flags later.

**Comment by zimbatm** [comment]
Sounds good. And if it was added to nixpkgs-review or a script, PR authors could get the feedback even earlier.

**Comment by infinisil** [comment]
> Before dismissing an idea, we can take a minute.

Sorry I was a bit unclear on that. I strongly believe that we should always have a good reason for making such changes. Imo ofborg automation and a greppable history are actually pretty good reasons, so the `<component>: <message>` format seems good to enforce.

> Also stylistic unification.

I'm really not sure whether that's good enough of a reason to yell at people when they use uppercase instead of lowercase :sweat_smile:. 

> I was thinking having a check for `<anything>: <anything>` would be a good first start. This should cover likely >90% of wrong commit titles.

That sounds pretty good to me :)

**Comment by zimbatm** [comment]
Ok, we have local consensus. @Atemu it's up to you to write a POC next.

**Comment by Atemu** [comment]
Note that, while I'm interested in implementing this, I likely won't have time to do so for at least a couple months. If anyone else wants to do this in the mean time, go ahead.

**Comment by zimbatm** [comment]
> Note that, while I'm interested in implementing this, I likely won't have time to do so for at least a couple months.

Ok. My impression was that you were motivated to take that on. Attention is a precious resource.

**Comment by Atemu** [comment]
> I would like to point out that CI has to get the logic for the commit messages from somewhere and if CI were pulling from a maintainers script in the nixpkgs repo itself, you may get a scenario where CI passes in the PR but noisily fails outside of it when the logic behind the CI is updated.

This is not a case that could happen in my proposal as only commits the PR adds atop of the target branch would be checked for conformance.

**Comment by misuzu** [comment]
> I was thinking having a check for `<anything>: <anything>` would be a good first start. This should cover likely >90% of wrong commit titles.

Seems like a good idea!

**Comment by sersorrel** [comment]
> Exceptional "components" that are always valid: `tree-wide:`

Note that `treewide:` is substantially more common; `git log --grep` produces 2043 commits containing the unhyphenated form, and only 87 commits containing the hyphenated form.

> Ability to name multiple components using commas (`pkg1, pkg2: fix foo`)

`git log --grep '{'` indicates that there is (also?) an existing pattern (~5564 commits, though that's definitely an overestimate) whereby [brace expansion syntax](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html) is imitated when making commits relating to multiple components, e.g. 86b5f88a2f39 ("{gcc8{,Stdenv},gfortran8}: drop"), debd6458f7cc ("{ponyc,pony-corral}: add numinit as maintainer"), d366b9a6f3b1 ("osu-lazer{,-bin}: add Guanran928 as maintainer").

I'd be wary of hard-enforcing much beyond "the first line matches `^.*\S: \S.*$` and the second line (if present) is blank" at least initially.

**Comment by jackrosenberg** [comment]
Just stumbled upon this while searching for the commit conventions/guidelines again. I think this is a good idea, it would be nice to have clear consistent checks. What's the latest on this?

**Comment by mdaniels5757** [comment]
I've done this (well, a somewhat more primitive version of it) in https://github.com/NixOS/nixpkgs/blob/38d9adc8caf6f834d22df4ed59afe9056e3e39a1/ci/github-script/lint-commits.js#L13.

**Comment by MattSturgeon** [comment]
> I've done this (well, a somewhat more primitive version of it) in `lint-commits.js`.

I think that's enough to consider this issue "completed"?
